gateway/src/metrics.rs

```diff
@@ -0,0 +1,1048 @@
+                             a spawned task panicking should abort the whole \
+                             process...";
+                            "error" => %e,
+                        );
+                    }
+                }
+                // drain any remaining errors
+                joined = self.tasks.try_join_next();
+            }
+        }
+    }
+}
+
+impl Drop for PollerManager {
+    fn drop(&mut self) {
+        // This is why the `JoinSet` is a field on the `PollerManager` struct
+        // rather than a local variable in `async fn run()`!
+        self.tasks.abort_all();
+    }
+}
+
+impl SpPoller {
+    async fn run(
+        mut self,
+        poll_interval: Duration,
+        apictx: Arc<ServerContext>,
+    ) -> Result<(), SpCommsError> {
+        /// How long to wait when a SP is not present.
+        ///
+        /// I picked this arbitrarily. It would be much nicer if we could
+        /// instead recieve a notification when the discovery state changes...
+        const NO_SP_BACKOFF: Duration = Duration::from_secs(30);
+        let mut interval = tokio::time::interval(poll_interval);
+        let switch = &apictx.mgmt_switch;
+        let sp = switch.sp(self.spid)?;
+        let mut not_present_message_logged = false;
+        loop {
+            interval.tick().await;
+            slog::trace!(&self.log, "interval elapsed, polling SP...");
+
+            match self.poll(sp).await {
+                // No sense cluttering the ringbuffer with empty vecs...
+                Ok(samples) if samples.is_empty() => {
+                    slog::trace!(&self.log, "polled SP, no samples returned"; "num_samples" => 0usize);
+                }
+                Ok(samples) => {
+                    slog::trace!(&self.log, "polled SP successfully"; "num_samples" => samples.len());
+
+                    if let Err(_) = self.sample_tx.send(samples) {
+                        slog::info!(
+                            &self.log,
+                            "all sample receiver handles have been dropped! \
+                             presumably we are shutting down...";
+                        );
+                        return Ok(());
+                    }
+                }
+                //
+                Err(CommunicationError::NoSpDiscovered) => {
+                    if !not_present_message_logged {
+                        not_present_message_logged = true;
+                        slog::info!(
+                            &self.log,
+                            "our SP seems to not be present, waiting to see if it \
+                             appears..."
+                        );
+                    } else {
+                        slog::debug!(
+                            &self.log,
+                            "SP is still not there, checking again in a little bit."
+                        )
+                    }
+
+                    tokio::time::sleep(NO_SP_BACKOFF).await;
+                }
```

Comment:
I dislike this bit. What I would *really* like to do instead is change `LocationMap` to store each SP's discovery state in a `tokio::sync::watch` channel, so that we can wait to receive a notification when a SP is discovered that was not previously present, instead of checking every 30 seconds. But, that feels like a much bigger change to MGS, so I'd like to do it as a separate PR instead.

Reply:
...oh, it turns out the SP address already lives in a `watch`, perhaps we can just use that!